File: server/services/webhook.js
const RETRY_DELAY_MS = 2000;

function log(message) {
  console.log(`[webhook ${new Date().toISOString()}] ${message}`);
}

async function postWebhook(url, token, payload) {
  const res = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify(payload),
  });

  if (!res.ok) {
    throw new Error(`Webhook target responded with status ${res.status}`);
  }

  return res;
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// Fires the ticket-created webhook asynchronously. Does not block the caller —
// callers should not await this in the request/response path.
async function sendTicketCreatedWebhook(ticket) {
  const url = process.env.WEBHOOK_TARGET_URL;
  const token = process.env.WEBHOOK_AUTH_TOKEN;

  if (!url) {
    log("skipped — WEBHOOK_TARGET_URL is not configured");
    return;
  }

  const payload = {
    event: "ticket.created",
    ticket: {
      id: ticket.id,
      subject: ticket.subject,
      description: ticket.description,
      channel: ticket.channel,
      requester_email: ticket.requesterEmail,
      created_at: ticket.createdAt,
    },
  };

  try {
    await postWebhook(url, token, payload);
    log(`success — ticket #${ticket.id} delivered to ${url}`);
    return;
  } catch (err) {
    log(`failed (attempt 1) for ticket #${ticket.id} — ${err.message}, retrying in ${RETRY_DELAY_MS}ms`);
  }

  await sleep(RETRY_DELAY_MS);

  try {
    await postWebhook(url, token, payload);
    log(`success on retry — ticket #${ticket.id} delivered to ${url}`);
  } catch (err) {
    log(`failed (attempt 2, final) for ticket #${ticket.id} — ${err.message}`);
  }
}

module.exports = { sendTicketCreatedWebhook };

File: fastapi-app/app/services/webhook.py
import asyncio
from datetime import datetime, timezone

import httpx

from ..config import settings

RETRY_DELAY_SECONDS = 2


def _log(message: str) -> None:
    ts = datetime.now(timezone.utc).isoformat()
    print(f"[webhook {ts}] {message}")


async def _post_webhook(payload: dict) -> None:
    headers = {"Content-Type": "application/json"}
    if settings.webhook_auth_token:
        headers["Authorization"] = f"Bearer {settings.webhook_auth_token}"

    async with httpx.AsyncClient(timeout=10) as client:
        response = await client.post(settings.webhook_target_url, headers=headers, json=payload)
        response.raise_for_status()


async def send_ticket_created_webhook(ticket: dict) -> None:
    """Fire-and-forget: scheduled via BackgroundTasks so it never delays the
    API response. Retries once, 2 seconds after a failed attempt."""
    if not settings.webhook_target_url:
        _log("skipped — WEBHOOK_TARGET_URL is not configured")
        return

    payload = {
        "event": "ticket.created",
        "ticket": {
            "id": ticket["id"],
            "subject": ticket["subject"],
            "description": ticket["description"],
            "channel": ticket["channel"],
            "requester_email": ticket["requester_email"],
            "created_at": ticket["created_at"],
        },
    }

    try:
        await _post_webhook(payload)
        _log(f"success — ticket #{ticket['id']} delivered to {settings.webhook_target_url}")
        return
    except Exception as err:  # noqa: BLE001 — POC: log and retry regardless of failure kind
        _log(f"failed (attempt 1) for ticket #{ticket['id']} — {err}, retrying in {RETRY_DELAY_SECONDS}s")

    await asyncio.sleep(RETRY_DELAY_SECONDS)

    try:
        await _post_webhook(payload)
        _log(f"success on retry — ticket #{ticket['id']} delivered to {settings.webhook_target_url}")
    except Exception as err:  # noqa: BLE001
        _log(f"failed (attempt 2, final) for ticket #{ticket['id']} — {err}")
